### ISSUE TITLE
Max OSX worker stalls with maven build

### ISSUE THREAD

**Issue by agibsonccc** [actions/virtual-environments#3243]
**Description**  
Maven job stalls on osx.

**Area for Triage**:  

**Question, Bug, or Feature?**:  

Bug
**Virtual environments affected**  
- [ ] Ubuntu 16.04
- [ ] Ubuntu 18.04
- [ ] Ubuntu 20.04
- [X ] macOS 10.15
- [ ] macOS 11.0
- [ ] Windows Server 2016 R2
- [ ] Windows Server 2019

**Image version**

Image version where you are experiencing the issue.
20210419.2

**Expected behavior**  
A description of what you expected to happen.
The mvn job should finish.

**Actual behavior**  
A description of what is actually happening.
The job seems to stall. Playing with it, I've had the build inconsistently sometimes finish when I break up the maven modules being built. I'm not convinced it's a ram issue. I've capped the memory usage at  2g just to be sure.

**Repro steps**  
A description with steps to reproduce the issue. If your have a public example or repo to share,
please provide the link.

Edit: Update job log for clarity:
https://github.com/eclipse/deeplearning4j/runs/2406050560?check_suite_focus=true

**Comment by dmitry-shibanov** [comment]
Hello @agibsonccc. Thank you for your report. I think the issue can be related to `setup-java` and `crazy-max/ghaction-import-gpg@v1`. Both `setup-java` and `crazy-max/ghaction-import-gpg@v1` import the same gpg private key. In post job step both of them remove keys, that is why post job for setup-java fails, because it does not contain key any more. I think you should not use them together, because they start to conflict. Could you please import the gpg key only once in `setup-java` or `crazy-max/ghaction-import-gpg@v1`

**Comment by agibsonccc** [comment]
@dmitry-shibanov  thanks for commenting. The job failures (at least to me) seem decoupled from that. The gpg keys did not work properly originally on windows which is why it was added in. I can try removing it, but I've found that the builds were correlated to the maven-javadoc-plugin on maven. 
This could be related to OOM, but this only seems to happen on OSX.

I've removed that and the builds seem to work. For future readers of this issue, could you explain why that might be the case?

That might explain some other issues I've been having with gpg key clean up though, thanks!

**Comment by dmitry-shibanov** [comment]
Hello @agibsonccc. Thank you for the answer. The root cause of the issue that task try to remove gpg private key which does not exist any more. On the post step of setup-java you didn't have key that is why it failed. 
- You had two steps with setup `gpg plugin key`. The first in [setup-java](https://github.com/actions/setup-java/blob/0443912ed73df2ee8596160daba017daab1c8ca9/src/gpg.ts#L12) and the second in [crazy-max/ghaction-import-gpg](https://github.com/crazy-max/ghaction-import-gpg/blob/318d42bcdade5ece53efd540c33f35b18aeb9bb5/src/gpg.ts#L106). Both tasks after generation try to remove private key ([setup-java](https://github.com/actions/setup-java/blob/main/src/cleanup-java.ts) and [crazy-max/ghaction-import-gpg](https://github.com/crazy-max/ghaction-import-gpg/blob/318d42bcdade5ece53efd540c33f35b18aeb9bb5/src/gpg.ts#L156)) in their post job steps. 
- After post job is finished they do not have key any more, that is why step with post logic was failed in setup-java, because it tried to remove key but it did not contain key any more. 
- We had the similar [issue](https://github.com/actions/setup-java/issues/128) in setup-java repository.

**Comment by agibsonccc** [comment]
@dmitry-shibanov  yes I understand that. I linked a second build that just crashes with zero logs. The GPG plugin is not doing that. I've had other builds fail for that reason due to cleanup, but they don't fail in a non descript way. I raised an issue, because only github  internal folks know why this class of failure occurs. 

Could you clarify that the gpg plugin is what causes the failure, when my experiments for the crashing builds instead point to an out of memory/machine failure (which I'd expect to be a bit harder to debug)

**Comment by agibsonccc** [comment]
@dmitry-shibanov  on a separate note (on both windows and mac I see this), setup-java fails to import the key properly which led me to add this other plugin in the first place:
https://github.com/eclipse/deeplearning4j/runs/2443237503?check_suite_focus=true

I'm happy to raise a separate issue, but first I'd prefer to have my builds succeeding even if the cleanup fails.

**Comment by dmitry-shibanov** [comment]
Hello @agibsonccc. Sorry for the misunderstanding. Does your issue reproduce only once or constantly ? I think issue can be related to GitHub [incident](https://www.githubstatus.com/incidents/n6v2xymbsjts).

**Comment by dmitry-shibanov** [comment]
I m closing the issue. Please ping me, if you still experience with this issue. The incident has been already resolved.

**Comment by agibsonccc** [comment]
@dmitry-shibanov  I haven't really tried the same configuration with mac osx since I raised the issue. I just disabled the javadoc configuration on mac os completely and upload javadoc from other builds instead. I wouldn't assume this is done yet due to this being a repeated issue. I would consider instead allowing a bit more transparency in to when a VM is shut down. I know it's not an easy problem, but that's the real only fix here I feel. "Your VM ran out of memory", "You exceeded the quota" still is not very transparent. I don't mind closing the issue as I don't intend on reivisting this for a while. I'll retry in the future and see if the changes improve. Keeping my builds where 1 build can't publish javadoc for  a release feels fragile at best.
